# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/jit/IonBuilder.cpp
# Commit: 6ccaefced3df
# Full Hash: 6ccaefced3df63b297057b31a856527153801a71
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2020-01-13 15:45:24
# Description:
#   Bug 1605641 - Mark 'done' stack value for destructuring assignment as having implicit uses. r=tcampbell
#   
#   We need to make sure this value isn't optimized out because the exception
#   handler may need it.
#   
# ==============================================================================

diff -r 96753184f3c4 -r 6ccaefced3df js/src/jit/IonBuilder.cpp
--- a/js/src/jit/IonBuilder.cpp	Thu Jan 02 12:07:51 2020 +0000
+++ b/js/src/jit/IonBuilder.cpp	Mon Jan 13 09:23:17 2020 +0000
@@ -7836,6 +7836,26 @@
     }
   }
 
+  // The bytecode emitted for destructuring assignments uses a "done" stack
+  // value that can be read from the exception handler. Mark the loop phi for
+  // this slot as having implicit uses so it won't be optimized away (replaced
+  // by the JS_OPTIMIZED_OUT magic value).
+  // See ProcessTryNotes in vm/Interpreter.cpp.
+  MOZ_ASSERT(block->stackDepth() >= info().firstStackSlot());
+  bool emptyStack = block->stackDepth() == info().firstStackSlot();
+  if (!emptyStack) {
+    JSContext* cx = TlsContext.get();
+    for (TryNoteIterAll tni(cx, script(), pc); !tni.done(); ++tni) {
+      const JSTryNote& tn = **tni;
+      if (tn.kind != JSTRY_DESTRUCTURING) {
+        continue;
+      }
+      MOZ_ASSERT(tn.stackDepth > 1);
+      uint32_t slot = info().stackSlot(tn.stackDepth - 1);
+      block->getSlot(slot)->setImplicitlyUsedUnchecked();
+    }
+  }
+
   return block;
 }
 